In [1]:
import os
os.getcwd()

'C:\\Users\\enriq\\OneDrive\\Desktop\\proyecto_olist'

## Data Loading

In this step, we load the main orders dataset.  
This dataset contains information about each purchase made.

In [2]:
import pandas as pd

orders = pd.read_csv(r"C:\Users\enriq\OneDrive\Desktop\proyecto_olist\olist_orders_dataset.csv")

orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


## Dataset Overview

In this step, we review the structure of the orders dataset to understand:
- the number of rows and columns
- the column names
- the data types

In [3]:
orders.shape

(99441, 8)

## Data Types and Structure

In this step, we examine the data types and non-null values of each column to understand data quality and structure.

In [4]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


## Understanding the Dataset Columns

In this step, we review the meaning of each column in the orders dataset to understand what each variable represents from a business perspective.

In [5]:
orders.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='object')

## Business Understanding of Columns

- **order_id**: Unique identifier for each order  
- **customer_id**: Unique identifier for each customer  
- **order_status**: Status of the order (delivered, shipped, canceled, etc.)  
- **order_purchase_timestamp**: Date and time when the order was placed  
- **order_approved_at**: Date and time when the payment was approved  
- **order_delivered_carrier_date**: Date when the order was handed to the carrier  
- **order_delivered_customer_date**: Date when the order was delivered to the customer  
- **order_estimated_delivery_date**: Estimated delivery date

## Delivered Orders Analysis

In this step, we analyze how many orders were successfully delivered.

This helps us understand operational performance and fulfillment success.

In [6]:
orders['order_status'].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

## Business Insight - Order Status

The majority of orders are in the "delivered" status, indicating a strong fulfillment performance.

A smaller portion of orders are canceled or unavailable, which could represent potential operational or customer experience issues.

Further analysis could focus on understanding the reasons behind cancellations.

## Delivery Rate

In this step, we calculate the percentage of orders that were successfully delivered.

This metric helps evaluate operational efficiency and customer fulfillment.

In [7]:
delivery_rate = (orders['order_status'] == 'delivered').mean()

delivery_rate

np.float64(0.9702034372140264)

## Business Insight - Delivery Rate

Approximately 97% of orders have been successfully delivered, indicating strong operational performance.

However, this metric may improve as orders currently in "shipped" status are completed.

There are also 609 unavailable orders, along with canceled and in-progress orders (such as invoiced), which require further analysis to understand their final outcomes.

A deeper investigation is needed to accurately assess total performance and identify potential inefficiencies in the fulfillment process.

## Delivery Time Analysis

In this step, we calculate the time it takes for orders to be delivered.

This metric is critical for understanding customer experience and logistics performance.

In [8]:
# convert to datetime
orders['order_purchase_timestamp'] = pd.to_datetime(orders['order_purchase_timestamp'])
orders['order_delivered_customer_date'] = pd.to_datetime(orders['order_delivered_customer_date'])

# calculate delivery time in days
orders['delivery_time_days'] = (
    orders['order_delivered_customer_date'] - orders['order_purchase_timestamp']
).dt.days

orders[['delivery_time_days']].head()

,delivery_time_days
0,8.0
1,13.0
2,9.0
3,13.0
4,2.0


## Average Delivery Time

In this step, we calculate the average number of days it takes for an order to be delivered.

This metric helps evaluate logistics efficiency and customer experience.

In [9]:
orders['delivery_time_days'].mean()

np.float64(12.094085575687217)

## Business Insight - Average Delivery Time

The average delivery time is approximately 12 days.

This indicates the typical waiting time customers experience after placing an order.

A delivery time of this length may impact customer satisfaction, especially in competitive e-commerce markets where faster shipping is expected.

Further analysis could explore:
- delivery time by region
- delays compared to estimated delivery date
- impact of delivery time on repeat purchases

## Delivery Performance vs Estimated Time

In this step, we compare the actual delivery date with the estimated delivery date.

This allows us to evaluate whether the company is meeting its delivery promises.

In [10]:
# convert estimated delivery date to datetime
orders['order_estimated_delivery_date'] = pd.to_datetime(orders['order_estimated_delivery_date'])

# calculate delay (actual - estimated)
orders['delivery_delay_days'] = (
    orders['order_delivered_customer_date'] - orders['order_estimated_delivery_date']
).dt.days

orders[['delivery_delay_days']].head()

,delivery_delay_days
0,-8.0
1,-6.0
2,-18.0
3,-13.0
4,-10.0


## Delivery Delay Distribution

In this step, we analyze the distribution of delivery delays to understand whether orders are typically delivered early, on time, or late.

In [11]:
orders['delivery_delay_days'].describe()

count    96476.000000
mean       -11.876881
std         10.183854
min       -147.000000
25%        -17.000000
50%        -12.000000
75%         -7.000000
max        188.000000
Name: delivery_delay_days, dtype: float64

## Business Insight - Delivery Performance

On average, orders are delivered approximately 12 days earlier than the estimated delivery date, suggesting conservative delivery estimates and strong logistics performance.

However, there are extreme outliers, with some orders delayed by up to 188 days.

These extreme delays, although potentially rare, could significantly impact customer satisfaction and should be investigated further.

A deeper analysis is required to:
- quantify how many delayed cases exist
- assess their frequency and impact
- identify root causes (e.g., logistics issues, location, product type)

## Late Deliveries Analysis

In this step, we calculate how many orders were delivered later than the estimated delivery date.

This helps quantify the scale of delivery issues.

In [12]:
late_orders = (orders['delivery_delay_days'] > 0).sum()

late_orders

np.int64(6535)

## Late Delivery Rate

In this step, we calculate the percentage of orders that were delivered late.

This metric helps quantify the impact of delivery delays on overall performance.

In [13]:
late_rate = (orders['delivery_delay_days'] > 0).mean()

late_rate

np.float64(0.06571736004263835)

## Business Insight - Late Delivery Rate

Approximately 6.6% of orders are delivered later than the estimated delivery date.

While the majority of deliveries arrive earlier than expected, this segment of late deliveries represents a critical risk for customer satisfaction and retention.

Even a relatively small percentage of delays can have a disproportionate impact on customer experience, especially if delays are significant.

Further analysis should focus on:
- identifying patterns in delayed orders (e.g., location, product category)
- evaluating the severity of delays
- understanding root causes to reduce future occurrences

## Delay Severity Analysis

In this step, we analyze how severe the delivery delays are by focusing only on late orders.

This helps differentiate between minor delays and critical failures in the delivery process.

In [14]:
late_delays = orders[orders['delivery_delay_days'] > 0]['delivery_delay_days']

late_delays.describe()

count    6535.000000
mean       10.620352
std        14.643844
min         1.000000
25%         3.000000
50%         7.000000
75%        13.000000
max       188.000000
Name: delivery_delay_days, dtype: float64

## Business Insight - Delay Severity

Among late deliveries, the average delay is approximately 10 days, indicating that delays are not minor but operationally significant.

The high standard deviation suggests substantial variability in delays, meaning that while some orders are only slightly delayed, others experience severe disruptions.

The 50th and 75th percentiles confirm that a large portion of delayed orders face considerable waiting times, which can negatively impact customer satisfaction.

Additionally, the presence of extreme outliers (maximum delay) indicates potential systemic issues or exceptional cases that require deeper investigation.

Overall, delivery delays are not only present but also severe enough to represent a meaningful risk to customer experience and operational reliability.

In [20]:
## Delivery Performance Segmentation

In this step, we classify orders into three categories:
- Early deliveries
- On-time deliveries
- Late deliveries

This helps us better understand overall delivery performance.

SyntaxError: invalid syntax (1415481088.py, line 3)

In [21]:
def classify_delivery(x):
    if pd.isna(x):
        return 'Not Delivered'
    elif x < 0:
        return 'Early'
    elif x == 0:
        return 'On Time'
    else:
        return 'Late'

orders['delivery_status'] = orders['delivery_delay_days'].apply(classify_delivery)

orders['delivery_status'].value_counts()

delivery_status
Early            88649
Late              6535
Not Delivered     2965
On Time           1292
Name: count, dtype: int64

## Business Insight - Delivery Performance Segmentation

The majority of orders (approximately 88,649) are delivered earlier than expected, indicating that delivery estimates are highly conservative and operations are generally efficient.

However, 6,535 orders are delivered late, representing a meaningful segment that may negatively impact customer satisfaction.

Additionally, 2,965 orders have not been delivered, which could indicate cancellations, logistics failures, or ongoing orders that require further investigation.

Only a small portion of orders are delivered exactly on time, suggesting that estimated delivery dates may not be optimized and could be adjusted to better reflect actual performance.

Overall, while early deliveries dominate, the presence of late and non-delivered orders highlights opportunities to improve reliability and customer experience.

## Relationship Between Delivery Time and Delays

In this step, we analyze whether longer delivery times are associated with higher delays.

This helps identify if slower deliveries are a key factor behind late orders.

In [16]:
orders[['delivery_time_days', 'delivery_delay_days']].corr()

,delivery_time_days,delivery_delay_days
delivery_time_days,1.000000,0.607716
delivery_delay_days,0.607716,1.000000


## Business Insight - Delivery Time vs Delays

There is a moderate to strong positive correlation (≈0.61) between delivery time and delivery delays.

This indicates that orders with longer delivery times are more likely to be delayed.

This suggests that delays are not random, but are associated with slower logistics processes.

Improving overall delivery speed could significantly reduce the number of late deliveries and enhance customer satisfaction.

## Comparing Late vs Non-Late Orders

In this step, we compare the average delivery time between late and non-late orders.

This helps determine whether delayed orders are significantly slower and quantify the impact.

In [17]:
orders.groupby(orders['delivery_delay_days'] > 0)['delivery_time_days'].mean()

delivery_delay_days
False    10.539665
True     33.487529
Name: delivery_time_days, dtype: float64

## Business Insight - Late vs Non-Late Orders

Late orders take significantly longer to be delivered, with an average delivery time of approximately 33.5 days compared to 10.5 days for non-late orders.

This indicates that delays are not random but are associated with structurally longer and potentially more complex deliveries.

The issue appears to be concentrated in a specific subset of orders rather than affecting the entire system.

This suggests potential challenges such as:
- long-distance deliveries
- complex logistics routes
- weaker estimation accuracy for longer delivery times

Targeting these high-risk orders could significantly reduce late deliveries and improve overall performance.

## Business Impact of Late Deliveries

In this step, we evaluate how significant late deliveries are in terms of volume.

This helps determine whether the issue is widespread or concentrated in a smaller segment of orders.

In [22]:
orders['delivery_status'].value_counts(normalize=True)

delivery_status
Early            0.891473
Late             0.065717
Not Delivered    0.029817
On Time          0.012993
Name: proportion, dtype: float64

## Business Insight - Impact of Late Deliveries

Late deliveries represent approximately 6.6% of total orders, indicating that the issue is relatively small in terms of volume.

However, these orders experience significantly longer delivery times and substantial delays, making them disproportionately impactful on customer satisfaction.

This suggests that while overall performance appears strong, a specific subset of problematic orders may be driving negative customer experiences.

Focusing on reducing delays within this segment could lead to meaningful improvements in overall service quality without requiring large-scale operational changes.

## Revenue Data Integration

In this step, we load the order items dataset to access price information and connect delivery performance with revenue.

This allows us to evaluate the financial impact of late deliveries.

In [23]:
order_items = pd.read_csv(r"C:\Users\enriq\OneDrive\Desktop\proyecto_olist\olist_order_items_dataset.csv")

order_items.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


## Merging Orders with Revenue Data

In this step, we merge the orders dataset with the order items dataset to associate each order with its corresponding revenue.

This allows us to analyze the financial impact of delivery performance.

In [24]:
df = orders.merge(order_items, on='order_id', how='left')

df.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,delivery_time_days,delivery_delay_days,delivery_status,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,8.0,-8.0,Early,1.0,87285b34884572647811a353c7ac498a,3504c0cb71d7fa48d967e0e4c94d59d9,2017-10-06 11:07:15,29.99,8.72
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13,13.0,-6.0,Early,1.0,595fac2a385ac33a80bd5114aec74eb8,289cdb325fb7e7f891c38608bf9e0962,2018-07-30 03:24:27,118.70,22.76
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04,9.0,-18.0,Early,1.0,aa4383b373c6aca5d8797843e5594415,4869f7a5dfa277a7dca6462dcf3b52b2,2018-08-13 08:55:23,159.90,19.22
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15,13.0,-13.0,Early,1.0,d0b61bfb1de832b15ba9d266ca96e5b0,66922902710d126a0e7d26b0e3805106,2017-11-23 19:45:59,45.00,27.20
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26,2.0,-10.0,Early,1.0,65266b2da20d04dbe00c5c2d3bb7859e,2c9e548be18521d1c43cde1c582c6de8,2018-02-19 20:31:37,19.90,8.72


## Revenue by Delivery Status

In this step, we calculate the total revenue associated with each delivery category.

This helps us understand how much money is affected by late deliveries.

In [25]:
df.groupby('delivery_status')['price'].sum().sort_values(ascending=False)

delivery_status
Early            12061944.61
Late               985954.33
Not Delivered      370748.59
On Time            172996.17
Name: price, dtype: float64

## Business Insight - Revenue Impact of Late Deliveries

Late deliveries account for approximately 7–8% of total revenue, representing a significant portion of the business.

Although this segment is relatively small in volume, it involves nearly 1 million in revenue, making it financially meaningful.

These delayed orders are likely to have a disproportionate impact on customer satisfaction, retention, and brand perception.

This suggests that improving delivery performance in this segment could protect high-value revenue and enhance overall business outcomes.

## Business Recommendations

Based on the analysis, the following actions are recommended:

1. Improve delivery time estimation for long-distance orders  
   - Orders with longer delivery times are more likely to be delayed  
   - Adjusting estimated delivery dates based on historical data can improve accuracy  

2. Focus on high-risk orders  
   - Late deliveries are concentrated in a specific subset of orders  
   - Identifying these orders early can help apply preventive actions  

3. Reduce extreme delays  
   - Some orders show very high delays (outliers)  
   - Investigating these cases can reveal operational issues or inefficiencies  

4. Optimize logistics for complex deliveries  
   - Long delivery times suggest possible routing or fulfillment inefficiencies  
   - Improving logistics in these cases could reduce delays significantly  

5. Monitor non-delivered orders  
   - Nearly 3% of orders are not delivered  
   - Understanding the causes can reduce lost revenue and improve reliability  

6. Prioritize customer experience for delayed orders  
   - Late deliveries represent ~7–8% of revenue  
   - Proactive communication or compensation strategies could reduce customer dissatisfaction